### PR TITLE
fix: Dataset version should allow catalog collection only

### PR DESCRIPTION
**Pull request by AmrouEmad**
## Reference

[Code review checklist](https://github.com/linz/geostore/blob/master/CODING.md#Checklist)

### CODE REVIEW

File: geostore/check.py [hunk 1]
```diff
@@ -13,3 +13,5 @@ class Check(Enum):
     STAGING_ACCESS = "staging bucket access"
     UNKNOWN_CLIENT_ERROR = "unknown client error"
     UNKNOWN_MULTIHASH_ERROR = "unknown multihash error"
+    UPLOADED_ASSETS_SHOULD_BE_CATALOG_OR_COLLECTION_ERROR = "assets should be catalog or collection"
```
Comment (l0b0) [inline geostore/check.py:16]: An *asset* should never be a catalogue or collection - an asset is a non-metadata file like a JPEG or SVG. How about something like `INVALID_STAC_ROOT_TYPE`? The error message in `geostore/check_stac_metadata/utils.py` already expands on what that means.

The same issue applies to `InvalidAssetFileError`.

File: geostore/check_stac_metadata/utils.py [hunk 2]
```diff
@@ -69,6 +67,14 @@ def maybe_convert_relative_url_to_absolute(url_or_path: str, parent_url: str) ->
     return f"{dirname(parent_url)}/{url_or_path}"
 
 
+def is_url_start_with_s3(metadata_url: str) -> bool:
```
Comment (l0b0) [inline geostore/check_stac_metadata/utils.py:70]: How about `is_s3_url`?

**Review by l0b0 — COMMENTED** [review]
Good stuff, just have some naming issues. You'll need to change the PR title as well, to fit the conventional commit format.

File: geostore/check_stac_metadata/utils.py [hunk 3]
```diff
@@ -138,12 +143,34 @@ def run(self, metadata_url: str) -> None:
...
-                    "error": NO_ASSETS_FOUND_ERROR_MESSAGE,
+                    "error": Check.NO_ASSETS_FOUND_ERROR_MESSAGE,
                     GIT_COMMIT: get_param(ParameterName.GIT_COMMIT),
                 },
             )
             return
 
+        s3_response = self.get_object(metadata_url)
+        stac_type = self.get_s3_url_as_object_json(metadata_url, s3_response)[STAC_TYPE_KEY]
+        if not is_instance_of_catalog_or_collection(stac_type):
```
Comment (l0b0) [inline geostore/check_stac_metadata/utils.py:154]: There's no test of this branch.

File: geostore/check_stac_metadata/utils.py (hunk 2 above)
Comment (AmrouEmad) [reply geostore/check_stac_metadata/utils.py]: Updated. Thanks

File: geostore/check.py (hunk 1 above)
Comment (AmrouEmad) [reply geostore/check.py]: Couldn't agree more. Updated.

File: geostore/check.py [hunk 4]
```diff
@@ -13,3 +13,5 @@ class Check(Enum):
...
+    NO_ASSETS_FOUND_ERROR_MESSAGE = "No assets found in dataset"
```
[the remark this answers is not in the document]
Comment (AmrouEmad) [reply geostore/check.py]: Updated. Thanks.

File: geostore/check.py [hunk 5]
```diff
@@ -13,3 +13,5 @@ class Check(Enum):
     STAGING_ACCESS = "staging bucket access"
     UNKNOWN_CLIENT_ERROR = "unknown client error"
     UNKNOWN_MULTIHASH_ERROR = "unknown multihash error"
+    INVALID_STAC_ROOT_TYPE = "Root type should be catalog or collection"
```
Comment (l0b0) [inline geostore/check.py:16]: The value should be formatted like the other ones, so it shouldn't be capitalised. By the way, can you please re-sort the enums so they're still alphabetical?

File: geostore/check_stac_metadata/utils.py (hunk 3 above)
Comment (AmrouEmad) [reply geostore/check_stac_metadata/utils.py:157]: Updated. Thanks

File: geostore/check_stac_metadata/utils.py [hunk 6]
```diff
@@ -69,6 +67,18 @@ def maybe_convert_relative_url_to_absolute(url_or_path: str, parent_url: str) ->
     return f"{dirname(parent_url)}/{url_or_path}"
 
 
+def is_s3_url(metadata_url: str) -> bool:
+    return metadata_url[:5] == S3_URL_PREFIX
+
+
+def is_empty(asset_list: List[Dict[str, str]]) -> int:
+    return len(asset_list)
```
Comment (l0b0) [inline geostore/check_stac_metadata/utils.py:75]: I don't understand this. Why is this called `is_empty`? And why is `len(asset_list)` pulled out as a function?

File: geostore/check.py (hunk 5 above)
Comment (l0b0) [reply geostore/check.py]: The list isn't quite sorted yet; `NO_ASSETS_IN_DATASET` should be in the middle. (This is why I recommend leaving the original commenter to resolve conversations - it ensures that each comment is verified as addressed by the commenter.)

File: geostore/check_stac_metadata/utils.py [hunk 7]
```diff
@@ -138,15 +147,41 @@ def run(self, metadata_url: str) -> None:
...
+                    "error": Check.NO_ASSETS_IN_DATASET,
                     GIT_COMMIT: get_param(ParameterName.GIT_COMMIT),
                 },
             )
             return
 
+        stac_type = self.get_stac_type_by_url(metadata_url)
+        if not is_instance_of_catalog_or_collection(stac_type):
+            error_message = (
+                f"Uploaded Assets should be catalog.json or collection.json”: “{metadata_url}”"
```
Comment (l0b0) [inline geostore/check_stac_metadata/utils.py:159]: The first "”" probably shouldn't be there.

File: geostore/check_stac_metadata/utils.py [hunk 8]
```diff
@@ -251,6 +277,19 @@ def validate(self, url: str) -> None:  # pylint: disable=too-complex
...
+            object_json: JsonObject = load(
+                s3_response.response,
+                object_pairs_hook=self.duplicate_object_names_report_builder(url),
+            )
+        except JSONDecodeError as error:
+            self.validation_result_factory.save(
+                url, Check.JSON_PARSE, ValidationResult.FAILED, details={MESSAGE_KEY: str(error)}
+            )
+            raise
+        return object_json
```
Comment (l0b0) [inline geostore/check_stac_metadata/utils.py:291]: The `return` can just be in the `try` block - it doesn't change the semantics of the code.

File: tests/test_check_stac_metadata_utils_logging.py [hunk 9]
```diff
@@ -101,9 +103,10 @@ def should_log_non_s3_url_prefix_validation() -> None:
     with patch("geostore.check_stac_metadata.utils.LOGGER.error") as logger_mock, patch(
         "geostore.check_stac_metadata.utils.processing_assets_model_with_meta"
     ):
-        STACDatasetValidator(
-            hash_key, url_reader, MockAssetGarbageCollector(), MockValidationResultFactory()
-        ).run(metadata_url)
+        with raises(InvalidSTACRootTypeError):
```
Comment (l0b0) [inline tests/test_check_stac_metadata_utils_logging.py:106]: This `with` could just be part of the parent `with`, as in `should_raise_invalid_stack_root_type_error_for_non_collection_or_catalog`.

**Comment by l0b0** [comment]
Also, the PR title still needs to change.

File: geostore/check_stac_metadata/utils.py (hunk 6 above)
Comment (AmrouEmad) [reply geostore/check_stac_metadata/utils.py:75]: You are right, but to pull functions out is more easier when mocking it.

File: geostore/check_stac_metadata/utils.py (hunk 7 above)
Comment (AmrouEmad) [reply geostore/check_stac_metadata/utils.py:159]: Thanks for spotting that. Updated

File: tests/test_check_stac_metadata_utils_logging.py (hunk 9 above)
Comment (AmrouEmad) [reply tests/test_check_stac_metadata_utils_logging.py:106]: Updated. Thanks.

File: geostore/check.py (hunk 5 above)
Comment (AmrouEmad) [reply geostore/check.py]: Updated. Thanks.

File: geostore/check_stac_metadata/utils.py (hunk 8 above)
Comment (AmrouEmad) [reply geostore/check_stac_metadata/utils.py]: Updated. Thanks.

**Pull request closed without merging**